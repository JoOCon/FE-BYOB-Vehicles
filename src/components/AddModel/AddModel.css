form {
  display: flex;
  flex-direction: column;
  width: 100%;
}

select, .add-make, input {
  width: 8rem;
}

input {
  display: block;
}