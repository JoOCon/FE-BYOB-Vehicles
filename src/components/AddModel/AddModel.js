import React, { Component } from 'react';
import './AddModel.css';

class AddModel extends Component {
  constructor() {
    super();
    this.state = {
      makes: [],
      toggleForm: false,
      makeName: '',
      manufacturer: '',
      makeId: 1,
      modelName: '',
      body: '',
      engine: '',
      topSpeed: '',
      hP: '',
      transmission: '',
    };
  }

  componentDidMount() {
    this.fetchMakes();
  }

  fetchMakes = () => {
    fetch(`${process.env.REACT_APP_DATABASE_API_URL}/api/v1/makes`)
      .then(response => response.json())
      .then(makes => this.setState({ makes }))
      .catch(error => console.log(error.message));
  }

  handleChange = (e) => {
    const { name, value } = e.target;
    this.setState({ [name]: value });
  }

  handleSubmitModel = (e) => {
    e.preventDefault()
    const {
      makeId,
      modelName,
      body,
      engine,
      topSpeed,
      hP,
      transmission,
    } = this.state;
    const newModel = {
      model_name: modelName,
      body,
      engine,
      top_speed: parseInt(topSpeed),
      horse_power: parseInt(hP),
      transmission,
      make_id: makeId
    };
    this.postNewModel(newModel);
    this.props.updateModels();
  }

  postNewModel = (model) => {
    fetch(`${process.env.REACT_APP_DATABASE_API_URL}/api/v1/models`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(model)
    });
  }

  handleSubmitMake = (e) => {
    e.preventDefault();
    const { makeName, manufacturer } = this.state;
    const newMake = { make_name: makeName, manufacturer };
    this.postNewMake(newMake);
    this.fetchMakes();
    this.setState({toggleForm: false, makeName: '', manufacturer: ''});
  }

  postNewMake = (make) => {
    fetch(`${process.env.REACT_APP_DATABASE_API_URL}/api/v1/makes`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(make)
    });
  }

  handleToggleForm = () => {
    const { toggleForm } = this.state
    this.setState({toggleForm: !toggleForm})
  }

  render() {
    const {
      toggleForm,
      makes,
      makesName,
      manufacturer,
      makeId,
      modelName,
      body,
      engine,
      topSpeed,
      hP,
      transmission,
    } = this.state;

    const listOptions = makes.map(make => (
      <option key={make.id} value={make.id}>{make.make_name}</option>
    ));

    const checkForModelInput = makeId === '' || modelName === '' || body === '' || engine === '' || topSpeed === '' || hP === '' || transmission === '';
    const checkForMakeInput = makesName === '' || manufacturer === '';

    const displayMakeForm = toggleForm ? 
      <form onSubmit={this.handleSubmitMake}>
        <input
          required
          type="text"
          placeholder="Make Name"
          name="makeName"
          value={this.state.makeName}
          onChange={this.handleChange}
        />
        <input
          required
          type="text"
          placeholder="manufacturer"
          name="manufacturer"
          value={this.state.manufacturer}
          onChange={this.handleChange}
        />
        <button disabled={checkForMakeInput}>Add Make</button>
      </form> :
      <div></div>

    return (
      <div>
        <select 
        name="makeId"
        onChange={this.handleChange}
        >
          {listOptions}
        </select>
        <button onClick={this.handleToggleForm}>Add Make</button>
        {displayMakeForm}
        <form onSubmit={this.handleSubmitModel}>
          <input
            required
            type="text"
            placeholder="Model Name"
            name="modelName"
            value={modelName}
            onChange={this.handleChange}
          />
          <input
            required
            type="text"
            placeholder="Body"
            name="body"
            value={body}
            onChange={this.handleChange}
          />
          <input
            required
            type="text"
            placeholder="Engine"
            name="engine"
            value={engine}
            onChange={this.handleChange}
          />
          <input
            required
            type="number"
            placeholder="Top Speed"
            name="topSpeed"
            value={topSpeed}
            onChange={this.handleChange}
          />
          <input
            required
            type="number"
            placeholder="HP"
            name="hP"
            value={hP}
            onChange={this.handleChange}
          />
          <input
            required
            type="text"
            placeholder="Transmission"
            name="transmission"
            value={transmission}
            onChange={this.handleChange}
          />
          <button disabled={checkForModelInput} className="add-model-btn">
            SUBMIT
          </button>
        </form>
      </div>
    );
  }
}

export default AddModel;
