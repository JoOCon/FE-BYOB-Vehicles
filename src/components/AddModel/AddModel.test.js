import React from 'react';
import { shallow } from 'enzyme';
import AddModel from './AddModel';

describe('AddModel', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<AddModel />);
  });
  it('should match snapshot with models', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('handleChange', () => {
    it('should set state when invoked', () => {
      const mockEvent = { target: { name: 'modelName', value: 'FF' } };
      wrapper.instance().handleChange(mockEvent);

      expect(wrapper.state().modelName).toBe('FF');
    });
  });
});
