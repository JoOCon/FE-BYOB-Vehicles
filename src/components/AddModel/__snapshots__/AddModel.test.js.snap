// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddModel should match snapshot with models 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AddModel />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <label
          htmlFor="makeName"
        >
          Make
        </label>,
        <select
          id="makeName"
          name="makeId"
          onChange={[Function]}
        />,
        <button
          className="add-make"
        >
          ADD MAKE
        </button>,
        <input
          name="modelName"
          onChange={[Function]}
          placeholder="Model Name"
          required={true}
          type="text"
          value=""
        />,
        <input
          name="body"
          onChange={[Function]}
          placeholder="Body"
          required={true}
          type="text"
          value=""
        />,
        <input
          name="engine"
          onChange={[Function]}
          placeholder="Engine"
          required={true}
          type="text"
          value=""
        />,
        <input
          name="topSpeed"
          onChange={[Function]}
          placeholder="Top Speed"
          required={true}
          type="number"
          value=""
        />,
        <input
          name="hP"
          onChange={[Function]}
          placeholder="HP"
          required={true}
          type="number"
          value=""
        />,
        <input
          name="transmission"
          onChange={[Function]}
          placeholder="Transmission"
          required={true}
          type="text"
          value=""
        />,
        <button
          className="add-model-btn"
          disabled={true}
        >
          SUBMIT
        </button>,
      ],
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Make",
          "htmlFor": "makeName",
        },
        "ref": null,
        "rendered": "Make",
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "id": "makeName",
          "name": "makeId",
          "onChange": [Function],
        },
        "ref": null,
        "rendered": Array [],
        "type": "select",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "ADD MAKE",
          "className": "add-make",
        },
        "ref": null,
        "rendered": "ADD MAKE",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "name": "modelName",
          "onChange": [Function],
          "placeholder": "Model Name",
          "required": true,
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "name": "body",
          "onChange": [Function],
          "placeholder": "Body",
          "required": true,
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "name": "engine",
          "onChange": [Function],
          "placeholder": "Engine",
          "required": true,
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "name": "topSpeed",
          "onChange": [Function],
          "placeholder": "Top Speed",
          "required": true,
          "type": "number",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "name": "hP",
          "onChange": [Function],
          "placeholder": "HP",
          "required": true,
          "type": "number",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "name": "transmission",
          "onChange": [Function],
          "placeholder": "Transmission",
          "required": true,
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "SUBMIT",
          "className": "add-model-btn",
          "disabled": true,
        },
        "ref": null,
        "rendered": "SUBMIT",
        "type": "button",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <label
            htmlFor="makeName"
          >
            Make
          </label>,
          <select
            id="makeName"
            name="makeId"
            onChange={[Function]}
          />,
          <button
            className="add-make"
          >
            ADD MAKE
          </button>,
          <input
            name="modelName"
            onChange={[Function]}
            placeholder="Model Name"
            required={true}
            type="text"
            value=""
          />,
          <input
            name="body"
            onChange={[Function]}
            placeholder="Body"
            required={true}
            type="text"
            value=""
          />,
          <input
            name="engine"
            onChange={[Function]}
            placeholder="Engine"
            required={true}
            type="text"
            value=""
          />,
          <input
            name="topSpeed"
            onChange={[Function]}
            placeholder="Top Speed"
            required={true}
            type="number"
            value=""
          />,
          <input
            name="hP"
            onChange={[Function]}
            placeholder="HP"
            required={true}
            type="number"
            value=""
          />,
          <input
            name="transmission"
            onChange={[Function]}
            placeholder="Transmission"
            required={true}
            type="text"
            value=""
          />,
          <button
            className="add-model-btn"
            disabled={true}
          >
            SUBMIT
          </button>,
        ],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Make",
            "htmlFor": "makeName",
          },
          "ref": null,
          "rendered": "Make",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "id": "makeName",
            "name": "makeId",
            "onChange": [Function],
          },
          "ref": null,
          "rendered": Array [],
          "type": "select",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "ADD MAKE",
            "className": "add-make",
          },
          "ref": null,
          "rendered": "ADD MAKE",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "name": "modelName",
            "onChange": [Function],
            "placeholder": "Model Name",
            "required": true,
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "name": "body",
            "onChange": [Function],
            "placeholder": "Body",
            "required": true,
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "name": "engine",
            "onChange": [Function],
            "placeholder": "Engine",
            "required": true,
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "name": "topSpeed",
            "onChange": [Function],
            "placeholder": "Top Speed",
            "required": true,
            "type": "number",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "name": "hP",
            "onChange": [Function],
            "placeholder": "HP",
            "required": true,
            "type": "number",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "name": "transmission",
            "onChange": [Function],
            "placeholder": "Transmission",
            "required": true,
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "SUBMIT",
            "className": "add-model-btn",
            "disabled": true,
          },
          "ref": null,
          "rendered": "SUBMIT",
          "type": "button",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
